Revamp of the LOAD CSV tutorial (#470)
neo4j/docs-getting-started: modules/ROOT/pages/data-import/csv-files.adoc

= Working with CSV files
:description: This page gives an overview of what is a dataset in CSV format and how to work with it when importing it into Neo4j.

When working with CSV files as the format of your data to be imported into Neo4j, there are a few points to be considered so that you can better use the available tools.
This page addresses the details of a CSV file and other recommendations when importing them.

== Structure of a CSV file

When analyzing the content of a CSV files, there are several elements that can help you understand the data that you are importing and xref:data-modeling/index.adoc[modeling] later on.

=== Data format

All data from the CSV file is read as a string, so you need to use `toInteger()`, `toFloat()`, `split()`, or similar functions to convert values, when needed.
Remember also that labels, property names, relationship types, and variables are *case-sensitive*.

See more information on how to manipulate data formats in the xref:#_converting_data_values[Converting data values] section.

=== Field terminator

Also known as delimiter, a field terminator is a character used to separate each field in a CSV file.

> In this example, a comma (`,`) is used, but other characters, such as a tab (`\`) or a pipe (`|`) also work and they can be blended:

Author: No, the tab is working normally here, both building locally and in Surge. Regarding the TSV file format, do you think it's better to mention it or to remove the tab option as it would make the file a TSV instead of a CSV, which is the topic of this page?

--
personId,name,birthYear
23945,"Pires, Gerard",1942
553509,"Reddy, Helen",1941
113934,"Flannery, Susan",1939
--

This CSV file uses commas as the xref:#_field_terminator[field terminator character] and, in the second row `names`, entries such as `Pires, Gerard` also contain a comma.
If the entry `Pires, Gerard` wasn't between quotation marks, it would count as two different entries (i.e. one for `Pires` and another for `Gerard`).

=== Normalized data

If the source data is normalized (e.g. when exported from a relational data model), there will typically be multiple CSV files.
Each CSV file will represent a table in the relational data model, and the files will be related to each other by unique IDs.

In this normalized data example, there are three files for people, movies, and roles:

.`person.csv`
[source,csv]
--
personId,name,birthYear
23945,Gerard Pires,1942
553509,Helen Reddy,1941
113934,Susan Flannery,1939
--

.`movies.csv`
[source,csv]
--
movieId,title,avgVote,releaseYear,genres
189,Sin City,8.000000,2005,Crime|Thriller
2300,The Fifth Element,7.700000,1997,Action|Adventure|Sci-Fi
11969,Tombstone,7.800000,1993,Action|Romance|Western
--

.`roles.csv`
[source,csv]
--

> personId,movieId,character

Author: I don't know if the people in the person.csv file are the actors or users, for example. But, at the same time, we're not going for this level of granularity here, so maybe it doesn't really matter?

If the source data is de-normalized, there will typically be a single CSV file which contains all the data, often duplicated where there are relationships between entities.
For example:

.`movies-n.csv`
[source,csv]
--
movieId,title,avgVote,releaseYear,genres,personType,name,birthYear,character
2300,The Fifth Element,7.700000,1997,Action|Adventure|Sci-Fi,ACTOR,Bruce Willis,1955,Korben Dallas
2300,The Fifth Element,7.700000,1997,Action|Adventure|Sci-Fi,ACTOR,Gary Oldman,1958,Jean-Baptiste Emanuel Zorg
2300,The Fifth Element,7.700000,1997,Action|Adventure|Sci-Fi,ACTOR,Ian Holm,1931,Father Vito Cornelius
11969,Tombstone,7.800000,1993,Action|Romance|Western,ACTOR,Kurt Russell,1951,Wyatt Earp
11969,Tombstone,7.800000,1993,Action|Romance|Western,ACTOR,Val Kilmer,1959,Doc Holliday
11969,Tombstone,7.800000,1993,Action|Romance|Western,ACTOR,Sam Elliott,1944,Virgil Earp
--

Here, the movie and person data (including the IDs) is repeated in different rows every time new information about a particular actor's role is featured.
This sort of duplication compromises the structure of the data, which means you need to xref:#_preparing_the_csv_file[prepare your file] before importing.

> == File location

Author: Maybe it's just the way it's phrased. I will tweak a bit.

When putting the data into a graph, there may not be a 1-1 mapping from the CSV file to a node or relationship.
+
To avoid this sort of problem:
+
* Check if headers match the data in the file.
* Adjust formatting, delimiters, columns, etc at this stage, so you don't need to spend more time later.

. *Extra or missing quotes*
+
Standalone double (`""`) or single quotes (`'`) in the middle of non-quoted text or non-escaped quotes in quoted text can cause issues when reading the file for loading.
It is best to either escape *or* remove stray quotes.
Find documentation for proper escaping in the link:https://neo4j.com/docs/cypher-manual/current/styleguide/#cypher-styleguide-meta-characters[Cypher style guide].

. *Special or newline characters*
+
When dealing with any special characters in a file, ensure they are quoted or remove them.
For newline characters in quoted or unquoted fields, either add quotes for these or remove them.

. *Inconsistent line breaks*
+
Computers do not handle well is inconsistent data.
Ensure line breaks are consistent throughout the file.

> The recommendation is to use the Unix style for compatibility with Linux systems (common format for import tools).

Author: I think this is a recommendation to make the file readable even in Linux systems, them being the exception, but I'm not sure.

To make the import process more efficient, you need to remove the unnecessary data _before_ you import the CSV files.

. *Inconsistency between headers and data*
+
Headers can be inconsistent with data.
They could be missing or be lost in too many columns.
+
To avoid this problem:
+
* Check if headers match the data in the file.
* Adjust formatting, columns, etc _before_ you import for a smooth process.

. *Extra or missing quotes*
+
Standalone double (`"`) or single quotes (`'`) in the middle of non-quoted text or non-escaped quotes in quoted text can cause issues when reading the file for loading.
It is best to either escape *or* remove stray quotes.
Find documentation for proper escaping in the link:https://neo4j.com/docs/cypher-manual/current/styleguide/#cypher-styleguide-meta-characters[Cypher style guide].

. *Special or newline characters*
+
When dealing with any special characters in a file, ensure they are quoted or remove them.
For newline characters in quoted or unquoted fields, either add quotes for these or remove them.

. *Inconsistent line breaks*
+
Ensure line breaks are consistent throughout the file.

> The recommendation is to use the Unix style for compatibility, in case you are using Linux.

Reviewer: ```suggestion
For Linux users, the recommendation is to use the Unix style for compatibility.
```

. *Binary zeros, BOM byte order mark (2 UTF-8 bytes) or other non-text characters*
+
Unusual characters or tool-specific formatting are sometimes hidden in application tools.
You can use a basic editor to detect and remove this type of characters from your files or link:https://neo4j.com/docs/cypher-manual/current/syntax/naming/#symbolic-names-escaping-rules[use backticks to escape them].

=== Data types

Since Neo4j reads all imported values as a string, you need to convert any values that are non-string.
You can do it using functions in Cypher:

* `*toInteger()*`: converts a value to an integer.
* `*toFloat()*`: converts a value to a float (e.g. for monetary amounts).

> * `*datetime()*`: converts a value to a `DateTime`.

Reviewer: For consistency, we should either use `code block` for all data types or none. I suggest to use it for all of them. So `string`, `float` etc.

== File size

You can use most Neo4j's xref:data-import/index.adoc#_methods_comparison[import methods] for importing small or medium-sized datasets (up to 10 million records).
If you want to import larger datasets, it is recommended to use link:https://neo4j.com/docs/operations-manual/current/import/[`neo4j-admin database import`].
See the tutorial for link:https://neo4j.com/docs/operations-manual/current/tutorial/neo4j-admin-import/[Neo4j-admin import] to learn more.

== Optimization

Performance can be a problem when working with large amounts of data or complex loading.
Some strategies can, however, improve the processing of large amounts of information.

For example, if you want to create a graph using xref:#_null_values[the preceding *companies.csv* file] and the following one:

.people.csv
[source]
----
employeeId,Name,companyId
1,Bob Smith,1
2,Joe Jones,3
3,Susan Scott,2
4,Karen White,1
----

In this case, you should separate node and relationship creation on a separate part of the processing.

> For instance, instead of the following:

Reviewer: ```suggestion
For example, instead of the following:
```
